File: [61] Rotate List/61.rotate-list.java
/*
 * @lc app=leetcode id=61 lang=java
 *
 * [61] Rotate List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) return head;
        // Calculate the length of the linked list
        int list_len = 0;
        ListNode dummy = new ListNode(0, head);
        while (head != null) {
            list_len++;
            head = head.next;
        }
        k = k%list_len;
        head = dummy.next;
        if (k == 0) return dummy.next;;
        ListNode current = head;
        while ((list_len - k) > 1) {
            current = current.next;
            k++;
        }
        dummy.next = current.next;
        current.next = null;
        current = dummy.next;
        while (current.next != null) current = current.next;
        current.next = head;
        return dummy.next;
    }
}
// @lc code=end

File: [61] Rotate List/61.rotate-list.cpp
/*
 * @lc app=leetcode id=61 lang=cpp
 *
 * [61] Rotate List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == nullptr || head->next == nullptr||k==0) return head;
        ListNode* node = head;
        int size = 1;  
        while(node->next != nullptr) {
            size++;
            node = node->next;
        }
        //handle the case of k>size
        k = k%size;
        if (k == 0) return head;
        //loop the list
        node->next=head;
        //find the node to break the loop at
        while(--size >= k) {
            node=node->next;
        }
        ListNode* first = node->next;
        node->next = nullptr;
        return first;
    }
};
// @lc code=end

